### PR TITLE
[IMP] tables: introduce dynamic tables

### PR DESCRIPTION
**Pull request by hokolomopo**
## [IMP] tables: introduce dynamic tables

This commit introduces the concept of dynamic table. A dynamic table
is a table that is bound to a array formula, instead of a static range.
The table will grow and shrink along the results of the array formula.

The notion is transparent to the user: if the users defines a table on
a single cell that contains an array formula, or all the cells of the
array formula, the table will be dynamic.

This means that the table zones are now an UI concept. The core plugin
now manages `InternalTable` objects, that can be either static or
dynamic (in which case they don't have a zone, but a single cell position).
The UI plugin `DynamicTablesPlugin` transform those `InternalTable` into
`Table` objects that every other plugin and component can easily use.

As the core plugin doesn't have access to the evaluation results, and
can't know if a cell contains an array formula, it's the responsability
of the components that create a table to determine if it's dynamic or not.

## [IMP] array formula: `getSpreadPositionsOf` now returns root position

Before this the getter `getSpreadPositionsOf` only returned the
position of the cells an array formula was spread to, minus the
cell the array formula was in. This made it impossible to differentiate
between a cell with a normal formula, and a cell with an array formula
returning a 1x1 matrix.

Now the getter returns the root position of the array formula as well.

There was also a bug where the spread relation weren't properly cleaned.

Task:  [3707037](https://www.odoo.com/web#id=3707037&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Review by anhe-odoo — COMMENTED** [review]
Some minor comments :)

File: src/helpers/ui/table_interactive.ts [hunk 1]
```diff
@@ -13,13 +14,21 @@ export function interactiveCreateTable(
...
   const selection = env.model.getters.getSelectedZones();
-  if (selection.length === 1 && getZoneArea(selection[0]) === 1) {
+  let target = env.model.getters.getSelectedZones();
+  const isDynamic = env.model.getters.canCreateDynamicTableOnZones(sheetId, selection);
+
+  if (selection.length === 1 && !isDynamic && getZoneArea(selection[0]) === 1) {
     env.model.selection.selectTableAroundSelection();
+    target = env.model.getters.getSelectedZones();
   }
-  const target = env.model.getters.getSelectedZones();
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/helpers/ui/table_interactive.ts:19]: 'cause I meant to use it in the `const isDynamic = ...` line instead of calling `env.model.getters.getSelectedZones()` twice :sweat:

File: src/plugins/core/tables.ts [hunk 2]
```diff
@@ -82,19 +76,17 @@ export class TablePlugin extends CorePlugin<TableState> implements TableState {
...
           return CommandResult.TableNotFound;
         }
         return this.checkValidations(cmd, this.checkUpdatedTableZoneIsValid, (cmd) =>
           this.checkTableConfigUpdateIsValid(cmd.config)
         );
       case "ADD_MERGE":
-        for (const merge of cmd.target) {
-          for (const table of this.getTables(cmd.sheetId)) {
-            if (overlap(table.range.zone, merge)) {
+        for (let merge of cmd.target) {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/tables.ts:87]: Rebase gone wrong for the let :smile:

File: src/plugins/core/tables.ts [hunk 3]
```diff
@@ -157,14 +152,31 @@ export class TablePlugin extends CorePlugin<TableState> implements TableState {
...
+              : this.updateStaticTable(table, range, newConfig);
+            this.history.update("tables", cmd.sheetId, table.id, newTable);
+          }
         }
         break;
       }
       case "UPDATE_CELL": {
         const sheetId = cmd.sheetId;
-        for (const table of this.getTables(sheetId)) {
+        for (let table of this.getInternalTables(sheetId)) {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/tables.ts:176]: rebase gone wrong :upside_down_face:

File: src/plugins/core/tables.ts [hunk 4]
```diff
@@ -175,10 +187,10 @@ export class TablePlugin extends CorePlugin<TableState> implements TableState {
         break;
       }
       case "DELETE_CONTENT": {
-        const tables: Record<TableId, Table | undefined> = { ...this.tables[cmd.sheetId] };
+        const tables: Record<TableId, InternalTable | undefined> = { ...this.tables[cmd.sheetId] };
         for (const tableId in tables) {
```
[the remark this answers is not in the document]
Comment (anhe-odoo) [reply src/plugins/core/tables.ts:192]: Ok, I didn't know about this :)

**Comment by rrahir** [comment]
THe first commit definitely deserves to be backported

File: src/plugins/core/tables.ts [hunk 5]
```diff
@@ -130,24 +125,25 @@ export class TablePlugin extends CorePlugin<TableState> implements TableState {
...
+          : this.createStaticTable(id, union, config);
         this.history.update("tables", cmd.sheetId, newTable.id, newTable);
         break;
       }
       case "REMOVE_TABLE": {
-        const tables: Record<UID, Table> = {};
-        for (const table of this.getTables(cmd.sheetId)) {
-          if (cmd.target.every((zone) => !intersection(zone, table.range.zone))) {
+        const tables: Record<UID, InternalTable> = {};
+        for (const table of this.getInternalTables(cmd.sheetId)) {
```
Comment (rrahir) [inline src/plugins/core/tables.ts:137]: It was already present before but this command can technically delete several tables at a time. should the command have been targeting the exact table zone definition all along?

Comment (hokolomopo) [reply src/plugins/core/tables.ts:137]: Probably could have been a good way to do it, but at the time it would have required a bit more logic in the menu items for no real reason.

Now I'm stuck with it, I don't see how I would migrate the revisions correctly in python.

**Review by LucasLefevre — COMMENTED** [review]
a few comments/suggestions :)

File: tests/evaluation/evaluation_formula_array.test.ts [hunk 6]
```diff
@@ -740,4 +743,38 @@ describe("evaluate formulas that return an array", () => {
...
+      ]);
+    });
+
+    test("getSpreadPositionsOf is correct after the evaluation changed so the formula can spread again", () => {
+      setCellContent(model, "H1", "5");
+      setCellContent(model, "A1", "=MFILL(H1,H1,42)");
+
+      expect(model.getters.getSpreadPositionsOf({ sheetId, col: 0, row: 0 })).toHaveLength(25);
+
+      setCellContent(model, "A4", "Block spread");
```
Comment (LucasLefevre) [inline tests/evaluation/evaluation_formula_array.test.ts:773]: I was running this test, trying to debug the evaluation cycle to asses what impacts your changes have. It turns out `"A1"` is now evaluated twice when setting content in "A4", even though it should not. One iteration with [A4, A1] (in that order), then a second iteration with A1 again (which is useless)
I'll check more in details what we should do instead tomorrow.

Comment (LucasLefevre) [reply tests/evaluation/evaluation_formula_array.test.ts:773]: If I understand well, the motivation behind the changes in the evaluation was for `getSpreadPositionsOf` (which was not including the top-left cell).
Could we just change the implementation of `getSpreadPositionsOf` to return what we want without changing anything else to the evaluation? Also, it could actually return a `Zone`.

File: src/types/table.ts [hunk 7]
```diff
@@ -27,6 +40,8 @@ export interface TableConfig {
   bandedRows: boolean;
   bandedColumns: boolean;
 
+  forceStaticTable?: boolean;
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/types/table.ts:43]: Changed to `"dynamic "| "static" | "forceStatic"`, and removed `forceStatic` option in the config

File: src/plugins/ui_core_views/dynamic_tables.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  private computeTables(sheetId: UID): Table[] {
+    const tables: Table[] = [];
+    const coreTables = this.getters.getCoreTables(sheetId);
+
+    // First we create the static tables, so we can use them to compute collision with dynamic tables
+    for (const table of coreTables) {
+      if (table.type === "dynamic") continue;
+      tables.push(this.coreTableToTable(sheetId, table));
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/dynamic_tables.ts:87]: call to `coreTableToTable` not useful here
```suggestion
      tables.push(table);
```

File: src/plugins/ui_core_views/dynamic_tables.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    return this.getTables(sheetId).filter((table) =>
+      zones.some((zone) => overlap(table.range.zone, zone))
+    );
+  }
+
+  doesZonesContainFilter(sheetId: UID, zones: Zone[]): boolean {
+    return this.getTablesOverlappingZones(sheetId, zones).some((table) => table.config.hasFilters);
+  }
+
+  doesZonesContainTable(sheetId: UID, zones: Zone[]): boolean {
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/dynamic_tables.ts:154]: not used anymore

File: src/plugins/ui_core_views/dynamic_tables.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  getFilterId(position: CellPosition): FilterId | undefined {
+    return this.getFilter(position)?.id;
+  }
+
+  getTable({ sheetId, col, row }: CellPosition): Table | undefined {
+    return this.getTables(sheetId).find((table) => isInside(col, row, table.range.zone));
+  }
+
+  /** Get the tables that are fully inside the given zone */
+  getTablesInZone(sheetId: UID, zone: Zone): Table[] {
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/dynamic_tables.ts:137]: not used anymore

File: tests/table/dynamic_table_plugin.test.ts [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+  Object.assign(TABLE_PRESETS, oldTablePresets);
+});
+
+function getTables(model: Model, sheetId: UID) {
+  return model.getters
+    .getTables(sheetId)
+    .map((table) => ({ ...table, zone: zoneToXc(table.range.zone) }));
+}
+
+describe("Dynamic tables", () => {
```
Comment (LucasLefevre) [inline tests/table/dynamic_table_plugin.test.ts:47]: add a test which duplicates a sheet with a dynamic table.  It doesn't seem there is one.

**Review by LucasLefevre — COMMENTED** [review]
a few more comments. I think we're good after :)

File: tests/table/dynamic_table_plugin.test.ts (hunk 11 above)
Comment (hokolomopo) [reply tests/table/dynamic_table_plugin.test.ts:47]: And it was, in fact, bugged :sweat_smile:

**Review by LucasLefevre — APPROVED** [review]
robodoo r+ rebase-ff
:)

**Comment by VincentSchippefilt** [comment]
robodoo r+ rebase-ff
